functional_tests.go

```diff
@@ -262,6 +262,59 @@ func testMakeBucketError() {
 	successLogger(function, args, startTime).Info()
 }
 
+func testMetadataSizeLimit() {
+	startTime := time.Now()
+	function := "PutObjectWithMetadata(bucketName, objectName, reader, metadata)"
+	args := map[string]interface{}{
+		"bucketName": "",
+		"objectName": "",
+		"metadata":   "",
+	}
+	rand.Seed(startTime.Unix())
+
+	// Instantiate new minio client object.
+	c, err := minio.New(
+		os.Getenv(serverEndpoint),
+		os.Getenv(accessKey),
+		os.Getenv(secretKey),
+		mustParseBool(os.Getenv(enableHTTPS)),
+	)
+	if err != nil {
+		failureLog(function, args, startTime, "", "Minio client creation failed", err).Fatal()
+	}
+	c.SetAppInfo("Minio-go-FunctionalTest", "0.1.0")
+
+	bucketName := randString(60, rand.NewSource(time.Now().UnixNano()), "minio-go-test")
+	args["bucketName"] = bucketName
+
+	objectName := randString(60, rand.NewSource(time.Now().UnixNano()), "")
+	args["objectName"] = objectName
+
+	err = c.MakeBucket(bucketName, "us-east-1")
+	if err != nil {
+		failureLog(function, args, startTime, "", "Make bucket failed", err).Fatal()
+	}
+
+	const HeaderSizeLimit = 8 * 1024
+	const UserMetadataLimit = 2 * 1024
+
+	metadata := make(map[string]string)
+	metadata["X-Amz-Meta-Mint-Test"] = string(bytes.Repeat([]byte("m"), 1+UserMetadataLimit-len("X-Amz-Meta-Mint-Test")))
+	args["metadata"] = fmt.Sprint(metadata)
+	_, err = c.PutObject(bucketName, objectName, bytes.NewReader(nil), 0, &minio.PutObjectOptions{UserMetadata: metadata})
+	if err == nil {
+		failureLog(function, args, startTime, "", "Created object with too much user-defined metadata successfully", nil).Fatal()
```

Comment:
Perhaps the message could be worded as
Created object with user-defined metadata exceeding metadata size limits.